Jakarta Data repositories in ServletContextListener

njr-11 opened the pull request:
Enables Jakarta Data repositories to be used within a ServletContextInitializer.
Adds tests of using ServletContextInitializer to pre-populate data with a Jakarta Data repository.

- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

njr-11 committed:
Issue #30879 disabled test for ServletContextListener scenario

njr-11 committed:
Issue #30879 support ServletContextListener

njr-11 committed:
Issue #30879 tests of ServletContextListener using a repository to initialize data

mswatosh commented on dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java:
```diff
@@ -601,6 +600,42 @@ private DataException excJNDINameNotFound(ComponentMetaData metadata,
 
+    /**
+     * Obtains the module name in which the repository interface is defined,
+     * or lacking that, the name of the application in which is is defined or used.
```
```suggestion
     * or lacking that, the name of the application in which it is defined or used.
```

njr-11 committed:
Update dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/cdi/FutureEMBuilder.java

Co-authored-by: Mark Swatosh <mark.swatosh@gmail.com>

njr-11 commented:
!build
notification=newcomment

njr-11 commented:
> https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=9d3fdbac-8e9d-4c98-88df-769e9521ebef - Open Liberty Personal Build completed with errors/failures.

The checkpoint test in data_fat fails intermittently with (3 out of 10) with,
```
[ERROR   ] CWWKC0453E: The server checkpoint request failed with the following message: CWWKC0457E: An error occurred preparing to take a checkpoint with the following message: CWWKZ0022W: Application DataTestApp has not started in 90.957 seconds.
```

njr-11 commented:
From the server log:
```
[3/27/25, 14:59:01:372 PDT] 0000003a id=58de4a31 io.openliberty.data.internal.persistence.DataProvider        3 componentMetaDataCreated DataTestApp#DataTestApp.war#test.jakarta.data.web.DataTestServlet 
                                                                                                               com.ibm.ws.webcontainer.osgi.metadata.WebComponentMetaDataImpl@105711c9[DataTestApp#DataTestApp.war#test.jakarta.data.web.DataTestServlet]
[3/27/25, 14:59:01:375 PDT] 0000003a id=00000000 com.ibm.ws.webcontainer.osgi.webapp.WebGroup                 I SRVE0169I: Loading Web Module: DataTestApp.
[3/27/25, 14:59:01:377 PDT] 0000003a id=00000000 com.ibm.ws.webcontainer.osgi.DynamicVirtualHost              I addWebApplication SRVE0250I: Web Module DataTestApp has been bound to default_host.
[3/27/25, 14:59:01:388 PDT] 0000003a id=00000000 SessionContextRegistryImpl                                   I getSessionContext SESN0176I: A new session context will be created for application key default_host/DataTestApp
[3/27/25, 14:59:01:393 PDT] 0000003a id=00000000 IDGeneratorImpl                                              I IDGeneratorImpl SESN0172I: The session manager is using the Java default SecureRandom implementation for session ID generation.
[3/27/25, 14:59:01:398 PDT] 0000003a id=00000000 SystemOut                                                    O DEBUG CheckpointPhase:Adding hook: com.ibm.ws.session.SessionContext$$Lambda$330/0x0000000000000000@42805e30 multi-threaded rank: 3
[3/27/25, 14:59:01:399 PDT] 0000003a id=00000000 SystemOut                                                    O DEBUG CheckpointPhase:Registering CheckpointHook service for rank :3 multi-threaded
[3/27/25, 14:59:01:399 PDT] 0000003a id=00000000 SystemOut                                                    O DEBUG CheckpointPhase:Added hook: true - 1
[3/27/25, 14:59:01:402 PDT] 0000003b id=00000000 IDGeneratorImpl                                              I IDGeneratorImpl SESN0172I: The session manager is using the Java default SecureRandom implementation for session ID generation.
[3/27/25, 14:59:01:403 PDT] 0000003b id=00000000 SystemOut                                                    O DEBUG CheckpointPhase:Adding hook: com.ibm.ws.session.SessionContext$$Lambda$330/0x0000000000000000@f63eb2c1 multi-threaded rank: 3
[3/27/25, 14:59:01:403 PDT] 0000003b id=00000000 SystemOut                                                    O DEBUG CheckpointPhase:Added hook: true - 2
[3/27/25, 14:59:01:477 PDT] 0000003a id=8c21b1b5 openliberty.data.internal.persistence.cdi.RepositoryProducer > produce Entry  
                                                                                                               org.jboss.weld.contexts.CreationalContextImpl@3f5ba148
                                                                                                               test.jakarta.data.web.Primes
[3/27/25, 14:59:01:487 PDT] 0000003a id=8c21b1b5 openliberty.data.internal.persistence.cdi.RepositoryProducer 3 add @jakarta.enterprise.concurrent.Asynchronous(executor="java:comp/DefaultManagedExecutorService", runAt={}) for public abstract java.util.concurrent.CompletableFuture test.jakarta.data.web.Primes.existsByRomanNumeralIgnoreCase(java.lang.String)
[3/27/25, 14:59:01:487 PDT] 0000003a id=8c21b1b5 openliberty.data.internal.persistence.cdi.RepositoryProducer 3 add @jakarta.enterprise.concurrent.Asynchronous(executor="java:comp/DefaultManagedExecutorService", runAt={}) for public abstract java.util.concurrent.CompletionStage test.jakarta.data.web.Primes.existsByNameIgnoreCase(java.lang.String)
[3/27/25, 14:59:01:488 PDT] 0000003a id=8c21b1b5 openliberty.data.internal.persistence.cdi.RepositoryProducer 3 add @jakarta.enterprise.concurrent.Asynchronous(executor="java:comp/DefaultManagedExecutorService", runAt={}) for public abstract java.util.concurrent.CompletionStage test.jakarta.data.web.Primes.findByNumberIdLessThanOrderByNumberIdDesc(long,jakarta.data.page.PageRequest)
[3/27/25, 14:59:01:488 PDT] 0000003a id=8c21b1b5 openliberty.data.internal.persistence.cdi.RepositoryProducer 3 add @jakarta.enterprise.concurrent.Asynchronous(executor="java:comp/DefaultManagedExecutorService", runAt={}) for public abstract java.util.concurrent.CompletableFuture test.jakarta.data.web.Primes.numberAsFloatWrapper(long)
[3/27/25, 14:59:01:488 PDT] 0000003a id=8c21b1b5 openliberty.data.internal.persistence.cdi.RepositoryProducer 3 add @jakarta.enterprise.concurrent.Asynchronous(executor="java:comp/DefaultManagedExecutorService", runAt={}) for public abstract java.util.concurrent.CompletableFuture test.jakarta.data.web.Primes.countByNumberIdBetweenAndEvenNot(long,long,boolean)
[3/27/25, 14:59:01:488 PDT] 0000003a id=8c21b1b5 openliberty.data.internal.persistence.cdi.RepositoryProducer 3 add @jakarta.enterprise.concurrent.Asynchronous(executor="java:comp/DefaultManagedExecutorService", runAt={}) for public abstract java.util.concurrent.CompletableFuture test.jakarta.data.web.Primes.divisibleByTwo(boolean,jakarta.data.page.PageRequest,jakarta.data.Order)
[3/27/25, 14:59:01:873 PDT] 0000003a id=00000000 io.openliberty.data.internal.persistence.RepositoryImpl      > <init> Entry  
                                                                                                               io.openliberty.data.internal.persistence.DataProvider@58de4a31
                                                                                                               io.openliberty.data.internal.persistence.cdi.DataExtension@9281b8d0
                                                                                                               FutureEMBuilder@637c8242(id=5b6a83bd) java:comp/DefaultDataSource DataTestApp#DataTestApp.war
                                                                                                               interface test.jakarta.data.web.Primes
                                                                                                               class test.jakarta.data.web.Prime
                                                                                                               {interface jakarta.data.repository.Query=[QueryInfo@54962613 java.util.List<java.lang.Long> minMaxSumCountAverageList(long) , ...

[3/27/25, 15:00:27:033 PDT] 0000003b id=00000000 com.ibm.ws.app.manager.AppMessageHelper                      A CWWKZ0022W: Application DataTestApp has not started in 90.000 seconds.
```

njr-11 committed:
ServletContextListener that accesses database is not compatible with Checkpoint

njr-11 commented:
Added a commit to disable the test that uses Jakarta Data from a ServletContextListener in a test bucket that uses Checkpoint/Restore because checkpoint is not compatible with a ServletContextListener accessing the database.

!build
notification=newcomment

njr-11 commented:
Also added a commit with the EJB workaround and to remove an incorrect approach of splitting into separate sets per module type.

!build
notification=newcomment

njr-11 committed:
Issue #30879 add workaround for EJB, remove incorrect approcah of splitting into separate sets, and enable the Singleton EJB test

njr-11 commented:
!build
notification=newcomment

njr-11 committed:
apply workaround to another place

njr-11 commented:
Updated to apply the temporary workaround in another place due to one remaining intermittent failure in the build.
!build
notification=newcomment
